codegen/smithy-aws-go-codegen/src/main/java/software/amazon/smithy/aws/go/codegen/AwsHttpPresignURLClientGenerator.java

```diff
@@ -166,138 +304,254 @@ public void writePresignClientType(GoWriter writer) {
+            ServiceShape serviceShape
+    ) {
+        writer.writeDocs(
+          String.format("%s represents presigner interface used by presign url client", presignerInterfaceSymbol.getName())
+        );
+        writer.openBlock("type $T interface {", "}", presignerInterfaceSymbol, () -> {
+            writer.write("PresignHTTP(");
+            writer.write("ctx context.Context, credentials aws.Credentials, r *http.Request, ");
+            writer.write("payloadHash string, service string, region string, signingTime time.Time, ");
+            writer.write(") (url string, signedHeader http.Header, err error)");
+        });
 
-        /**
-         * Sets the symbol provider for the builder
-         *
-         * @param symbolProvider the symbol provider
-         * @return buidler
-         */
-        public Builder symbolProvider(SymbolProvider symbolProvider) {
-            this.symbolProvider = symbolProvider;
-            return this;
-        }
+       writer.insertTrailingNewline();
+    }
 
-        /**
-         * Sets the operation for the builder
-         *
-         * @param operation api operation
-         * @return builder
-         */
-        public Builder operation(OperationShape operation) {
-            this.operation = operation;
-            return this;
-        }
+    /**
+     * Writes the Presign client's type and methods.
+     *
+     * @param writer writer to write to
+     */
+    public void writePresignOptionType(
+            GoWriter writer,
+            Model model,
+            SymbolProvider symbolProvider,
+            ServiceShape serviceShape
+    ) {
+        writer.addUseImports(SmithyGoDependency.CONTEXT);
+        Symbol presignOptionSymbol = buildSymbol(PRESIGN_OPTIONS, true);
+
+        // generate presign options
+        writer.writeDocs(String.format("%s represents the presign client options", PRESIGN_OPTIONS));
+        writer.openBlock("type $T struct {", "}", presignOptionSymbol, () -> {
+            writer.write("");
+            writer.writeDocs(
+                    "ClientOptions are list of functional options to mutate client options used by presign client"
+            );
+            writer.write("ClientOptions []func(*Options)");
+
+            writer.write("");
+            writer.writeDocs("Presigner is the  presigner used by the presign url client");
+            writer.write("Presigner $T", presignerInterfaceSymbol);
+
+            // s3 service has an additional Expires options
+            if (isS3ServiceShape(model, serviceShape)) {
+                writer.write("");
+                writer.writeDocs(
+                        String.format("Expires sets the expiration duration for the generated presign url. This should "
+                                + "be the duration in seconds the presigned URL should be considered valid for."
+                        )
+                );
+                writer.write("Expires time.Duration");
```

Comment:
Should call out the default value if unset. Since it seems like this header is required.